File: backend/internal/service/permission.go
package service

import (
	"backend/internal/model"
	"backend/internal/repository"
	"backend/pkg/casbin"
	"fmt"
)

// PermissionService 权限管理服务
type PermissionService struct {
	RoleRepo       *repository.RoleRepository
	MenuRepo       *repository.MenuRepository
	PermissionRepo *repository.PermissionRepository
}

// NewPermissionService 创建权限管理服务
func NewPermissionService(roleRepo *repository.RoleRepository, menuRepo *repository.MenuRepository, permissionRepo *repository.PermissionRepository) *PermissionService {
	return &PermissionService{
		RoleRepo:       roleRepo,
		MenuRepo:       menuRepo,
		PermissionRepo: permissionRepo,
	}
}

// CreatePermission 创建权限
func (s *PermissionService) CreatePermission(permission *model.Permission) error {
	return s.PermissionRepo.Create(permission)
}

// GetPermissionByID 根据ID获取权限
func (s *PermissionService) GetPermissionByID(id uint) (*model.Permission, error) {
	return s.PermissionRepo.GetByID(id)
}

// GetPermissions 获取权限列表
func (s *PermissionService) GetPermissions(limit, offset int, path, method string) ([]model.Permission, error) {
	return s.PermissionRepo.List(limit, offset, path, method)
}

// GetPermissionTotal 获取权限总数
func (s *PermissionService) GetPermissionTotal(path, method string) (int64, error) {
	return s.PermissionRepo.GetTotal(path, method)
}

// UpdatePermission 更新权限
func (s *PermissionService) UpdatePermission(permission *model.Permission) error {
	return s.PermissionRepo.Update(permission)
}

// UpdatePermissionStatus 更新权限状态
func (s *PermissionService) UpdatePermissionStatus(id uint, status int8) error {
	return s.PermissionRepo.UpdateStatus(id, status)
}

// GetAllPermissions 获取所有权限
func (s *PermissionService) GetAllPermissions(path, method string) ([]model.Permission, error) {
	return s.PermissionRepo.GetAll(path, method)
}

// DeletePermission 删除权限
func (s *PermissionService) DeletePermission(id uint) error {
	return s.PermissionRepo.Delete(id)
}

// // GetPermissionByPathAndMethod 根据路径和方法获取权限
// func (s *PermissionService) GetPermissionByPathAndMethod(path, method string) (*model.Permission, error) {
// 	return s.PermissionRepo.GetByPathAndMethod(path, method)
// }

// GetRoleMenus 获取角色的菜单权限
// func (s *PermissionService) GetRoleMenus(roleID uint) ([]model.Menu, error) {
// 	role, err := s.RoleRepo.GetByID(roleID)
// 	if err != nil {
// 		return nil, err
// 	}
// 	return role.Menus, nil
// }

// AddPolicy 添加Casbin策略
func (s *PermissionService) AddPolicy(roleID uint, path, method string) error {
	ident, err := s.RoleRepo.GetIdent(roleID)
	if err != nil {
		return err
	}
	if ident == "" {
		return fmt.Errorf("role identifier is missing")
	}
	sub := ident
	policy, err := casbin.Enforcer.AddPolicy(sub, path, method)
	fmt.Println("policy:", policy)
	if err != nil {
		return err
	}
	return nil
}

// RemovePolicy 移除Casbin策略
func (s *PermissionService) RemovePolicy(roleID uint, path, method string) error {
	ident, err := s.RoleRepo.GetIdent(roleID)
	if err != nil {
		return err
	}
	if ident == "" {
		return fmt.Errorf("role identifier is missing")
	}
	sub := ident
	policy, err := casbin.Enforcer.RemovePolicy(sub, path, method)
	fmt.Println("policy:", policy)
	if err != nil {
		return err
	}
	return nil
}

// GetPolicies 获取角色的所有Casbin策略
func (s *PermissionService) GetPolicies(roleID uint) ([][]string, error) {
	ident, err := s.RoleRepo.GetIdent(roleID)
	if err != nil {
		return nil, err
	}
	sub := ident
	return casbin.Enforcer.GetFilteredPolicy(0, sub)
}

// GetAllPolicies 获取所有Casbin策略
func (s *PermissionService) GetAllPolicies() ([][]string, error) {
	return casbin.Enforcer.GetPolicy()
}

File: frontend/src/api/permission.ts
import request from './request'

// 权限管理相关接口

// 为角色分配菜单权限
export const assignMenuToRole = (roleId: number, menuIds: number[]) => {
  return request.post(`/roles/${roleId}/menus`, { menu_ids: menuIds })
}

// 获取角色的菜单权限
export const getRoleMenus = (roleId: number) => {
  return request.get(`/roles/${roleId}/menus`)
}

// 移除角色的菜单权限
export const removeMenuFromRole = (roleId: number, menuIds: number[]) => {
  return request.delete(`/roles/${roleId}/menus`, { data: { menu_ids: menuIds } })
}

// 添加Casbin策略
export const addPolicy = (roleId: number, path: string, method: string) => {
  return request.post(`/roles/${roleId}/policies`, { path, method })
}

// 移除Casbin策略
export const removePolicy = (roleId: number, path: string, method: string) => {
  return request.delete(`/roles/${roleId}/policies`, { data: { path, method } })
}

// 获取角色的Casbin策略
export const getPolicies = (roleId: number) => {
  return request.get(`/roles/${roleId}/policies`)
}

// 获取所有Casbin策略
export const getAllPolicies = () => {
  return request.get('/policies')
}

// 权限资源管理相关接口

// 创建权限
export const createPermission = (data: { path: string; method: string; description?: string; status?: number }) => {
  return request.post('/permissions', data)
}

// 获取权限列表
export const getPermissions = (params?: { page?: number; page_size?: number; path?: string; method?: string }) => {
  return request.get('/permissions', { params })
}

// 获取权限详情
export const getPermission = (id: number) => {
  return request.get(`/permissions/${id}`)
}

// 根据路径和方法获取权限详情
export const getPermissionByPathAndMethod = (path: string, method: string) => {
  return request.get('/permissions/detail', { params: { path, method } })
}

// 更新权限
export const updatePermission = (id: number, data: { path?: string; method?: string; description?: string; status?: number }) => {
  return request.put(`/permissions/${id}`, data)
}

// 更新权限状态
export const updatePermissionStatus = (id: number, status: number) => {
  return request.put(`/permissions/${id}/status`, { status })
}

// 删除权限
export const deletePermission = (id: number) => {
  return request.delete(`/permissions/${id}`)
}